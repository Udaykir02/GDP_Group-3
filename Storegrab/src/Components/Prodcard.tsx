import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { StyleSheet, Dimensions, Image, TouchableWithoutFeedback, Alert, View, Text, Platform } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import { Card } from 'react-native-paper';
import { useAppTheme } from '../styles/theme/theme';
import { UserCartType } from '@/../reducers/users/types';
import Stepper from './shared/stepper/Stepper';
import { addItemToCart, decreaseItemQuantity, increaseItemQuantity, removeItemFromCart, updateItemInCart } from '../../reducers/cartReducer';
import { decreaseProductQuantity, increaseProductQuantity } from '../../reducers/vendorReducer';
import { addToCartRequest } from '../../actions/userActions';
import { decreaseInventoryQtyRequest, increaseInventoryQtyRequest } from '../../actions/vendorActions';
// import { Text, theme } from 'galio-framework';


// import NumericInput from '../components/NumericInput';
// import fire from '../firebase/Fire';

interface Product {
    title: string;
    meat: string;
    price: number;
    description: string;
    image: string;
    units: number;
    cart: any[];
}

interface Shop {
    cta: boolean;
}

interface Location { }

interface CartItem {
    uniqueID: string;
    cost: number;
}

interface ProdcardProps {
    navigation: any;
    item: Product;
    horizontal: boolean;
    full: boolean;
    style: any;
    ctaColor: string;
    imageStyle: any;
    shop: Shop;
    location: Location;
    cartitems: CartItem[];
}

const Prodcard: React.FC<any> = ({
    item,
    horizontal,
    style,
    imageStyle,
}) => {
    const [isHorizontal, setIsHorizontal] = useState<boolean>(horizontal);
    const { colors } = useAppTheme();
    const { user, token, vendorAdmin } = useSelector((state: any) => state.auth)
    const styles = StyleSheet.create({
        card: {
            backgroundColor: '#fff',
            marginVertical: 16,
            borderWidth: 0,
            minHeight: 114,
            marginBottom: 16
        },
        cardTitle: {
            flex: 1,
            flexWrap: 'wrap',
            paddingBottom: 6
        },
        cardDescription: {
            padding: 16 / 2
        },
        imageContainer: {
            borderRadius: 3,
            elevation: 1,
            overflow: 'hidden',
        },
        image: {
            // borderRadius: 3,
        },
        horizontalImage: {
            height: 107,
            width: 'auto',
            borderRadius: 3,
        },
        horizontalStyles: {
            borderTopRightRadius: 0,
            borderBottomRightRadius: 0,
        },
        verticalStyles: {
            borderBottomRightRadius: 0,
            borderBottomLeftRadius: 0
        },
        fullImage: {
            height: 170
        },
        shadow: {
            shadowColor: '#000',
            shadowOffset: { width: 0, height: 2 },
            shadowRadius: 4,
            shadowOpacity: 0.1,
            elevation: 2,
        },
        androidShadow: {
            borderWidth: 0,
            shadowColor: '#fff',
            shadowOffset: { width: 0, height: 0 },
            shadowRadius: 0,
            shadowOpacity: 0,
            elevation: 2, 
        },
        cardContainer: {
            padding: 2,
            backgroundColor: colors.backgroundDefault,
            elevation: 5,
            borderRadius: 8,
            marginVertical: 10
        },
    });
    const dispatch = useDispatch();

    const cartData = useSelector((state: any) => state.usercart.cart)



    const imageStyles = [
        !isHorizontal ? styles.fullImage : styles.horizontalImage,
        imageStyle,
    ];
    const cardContainer = [styles.card, style];
    const imgContainer = [
        styles.imageContainer,
        isHorizontal ? styles.horizontalStyles : styles.verticalStyles,
        Platform.OS === 'android'?  styles.androidShadow:styles.shadow,
    ];




    const handleChange = (change: number) => {
        if (vendorAdmin) {

            if (change > 0) {
                dispatch(increaseInventoryQtyRequest(item?.skuId, change, token))
                dispatch(increaseProductQuantity(item?.skuId))
            }
            else {
                if (item?.qty > 0) {
                    dispatch(decreaseInventoryQtyRequest(item?.skuId, 1, token))
                    dispatch(decreaseProductQuantity(item?.skuId))
                }

            }
        }
        else {
            let cartItem = cartData.find((cartItem: any) => cartItem.skuId === item?.skuId)
            dispatch(addToCartRequest(user?.userId, item?.skuId, change, token));
            if (cartItem) {
                if (cartItem.qty === 1 && change < 0) {
                    dispatch(removeItemFromCart(cartItem?.skuId))
                    dispatch(increaseProductQuantity(cartItem?.skuId))
                }
                else if (change > 0) {
                    if (item?.qty > 0) {
                        dispatch(increaseItemQuantity(cartItem?.skuId))
                        dispatch(decreaseProductQuantity(cartItem?.skuId))
                    }

                }
                else {
                    dispatch(decreaseItemQuantity(cartItem?.skuId))
                    dispatch(increaseProductQuantity(cartItem?.skuId))
                }
            }
            else {
                if (change > 0) {
                    if (item.qty > 0) {
                        const newCartItem: any = {
                            skuId: item?.skuId,
                            item: item?.item,
                            price: item?.price,
                            qty: change,
                            size: item?.size,
                            features: item?.features,
                            categories: item?.categories,
                            image: item?.image,
                            description: item?.description,
                            brand: item?.brand,
                            vendorId: item?.vendorId,
                            vendor_name: item?.vendor_name,
                        }
                        dispatch(addItemToCart(newCartItem))
                        dispatch(decreaseProductQuantity(cartItem?.skuId))
                    }

                }
            }
        }


    }

    const getQuantity = () => {
        if (vendorAdmin) {
            return item?.qty ? item?.qty : 0;
        }
        return (cartData && cartData.length > 0 && cartData.find((cartItem: any) => cartItem.skuId === item.skuId)?.qty) ? cartData.find((cartItem: any) => cartItem.skuId === item.skuId)?.qty : 0
    }

    return (
        <TouchableWithoutFeedback onPress={() => setIsHorizontal(!isHorizontal)}>
            <View style={[cardContainer, { flexDirection: isHorizontal ? 'row' : 'column' }]}>
                <View style={[imgContainer, { flex: isHorizontal ? 0.3 : 1 }]}>
                    <Image source={{ uri: item.image }} style={imageStyles} testID='prodcard-image-testid' />
                </View>
                <View style={{ flex: isHorizontal ? 0.7 : 1, justifyContent: 'space-between' }}>
                    <View style={{ justifyContent: 'space-between', flexDirection: 'row' }}>
                        <View style={[{ flex: 0.4 }, isHorizontal ? { marginHorizontal: 5 } : { marginVertical: 5 }]}>
                            <Text style={[styles.cardTitle, { color: "#32325D", fontWeight: 'bold', fontSize: 14 }]}>{item?.item}</Text>
                            <Text style={[styles.cardTitle, { color: "#32325D", fontWeight: 'normal', fontSize: 12 }]}>{item?.features}</Text>
                        </View>
                        <View style={{ flex: 0.6, alignItems: isHorizontal ? 'center' : 'flex-end', justifyContent: isHorizontal ? 'center' : 'flex-end' }}>
                            <Text style={[styles.cardTitle, { color: "#32325D", fontWeight: 'bold', fontSize: 18 }]}>{'\u0024' + item?.price}</Text>
                            {((!vendorAdmin && item?.qty > 0) || (cartData && cartData.length > 0 && cartData.find((cartItem: any) => cartItem.skuId === item.skuId)?.qty>0) || vendorAdmin)?<Stepper quantity={getQuantity()} handleChange={handleChange} />: <Text style={[styles.cardTitle, { color: "#32325D", fontWeight: 'bold', fontSize: 18 }]}>{'Out Of Stock'}</Text>}
                            <Text style={[styles.cardTitle, { color: "#32325D", fontWeight: 'bold', fontSize: 12, marginTop: 10 }]}>In Stock: {item?.qty}</Text>
                        </View>
                    </View>
                    <View style={{ justifyContent: 'space-between', flexDirection: 'row', margin: 5 }}>
                        <Text style={[styles.cardTitle, { color: "#32325D", fontWeight: 'bold', fontSize: 12 }]}>Description: {item?.description}</Text>
                    </View>
                    <View style={[{ justifyContent: 'space-between' }, isHorizontal ? { marginHorizontal: 5 } : { marginVertical: 5 }]}>
                        <Text style={[styles.cardTitle, { color: "#32325D", fontWeight: 'bold', fontSize: 12 }]}>{item.brand}</Text>
                    </View>
                </View>
            </View>
        </TouchableWithoutFeedback>

    );
};


export default Prodcard;