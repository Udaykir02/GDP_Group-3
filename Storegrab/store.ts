import { configureStore } from '@reduxjs/toolkit';
import createSagaMiddleware from 'redux-saga';
import rootReducers from './reducers'; // Assuming rootReducers is defined in a separate file
import rootSaga from './sagas/rootSaga'; // Assuming rootSaga is defined in a separate file

// Create saga middleware instance
const sagaMiddleware = createSagaMiddleware();

// Create the Redux store with middleware and reducers
const store = configureStore({
  reducer: rootReducers,
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(sagaMiddleware),
});

// Run the root saga
sagaMiddleware.run(rootSaga);

export default store;
