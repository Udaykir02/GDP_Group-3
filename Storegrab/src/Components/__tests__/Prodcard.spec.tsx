import { Provider } from 'react-redux'
import store from '../../../store'
import render, {
    RenderResult,
} from '@testing-library/react-native/build/render'
import { NavigationContainer } from '@react-navigation/native'
import { fireEvent } from '@testing-library/react-native';
import Prodcard from '../Prodcard';
import { StripeProvider } from '@stripe/stripe-react-native';
import { PaperProvider } from 'react-native-paper';
import AppThemeProvider from '../../core/AppThemeProvider';

jest.mock('@react-navigation/native', () => {
    return {
        ...jest.requireActual('@react-navigation/native'),
        useNavigation: () => ({
            navigate: jest.fn(),
        }),
    };
});

const inventoryMock = {
    skuId: 'sku_123',
    item: 'item',
    price: 100,
    qty: 1,
    size: {
      h: 1,
      l: 1,
      w: 1,
    },
    features: 'red',
    categories: ['category'],
    image: 'https://fastly.picsum.photos/id/237/200/300.jpg?hmac=TmmQSbShHz9CdQm0NkEjx1Dyh_Y984R9LpNrpvH2D_U',
    description:   'description',
    brand: 'brand',
    vendorId: 'vendorId',
    vendor_name: 'vendor_name',
  };
const renderStartupContainerPage = () => {
    return render(
            <Provider store={store}>
                <PaperProvider>
                    <AppThemeProvider>
                    <Prodcard
                            item={inventoryMock}
                            shop={{ cta: true }}
                            location={0}
                            horizontal
                            style={{ paddingHorizontal: 16, paddingVertical: 16 / 2 }}
                            cartitems={[]}
                            full={false} ctaColor={""} imageStyle={{ borderRadius: 10 }} />
                    </AppThemeProvider>
                </PaperProvider>
            </Provider>
    )
}
let page: RenderResult

describe('Prodcard', () => {
    beforeEach(() => {
        jest.useFakeTimers();
        page = renderStartupContainerPage()
    })
    afterEach(() => {
        jest.runOnlyPendingTimers();
        jest.useRealTimers();
    });
    it('should render correctly', function () {
        expect(page.toJSON()).toMatchSnapshot()
    })

    it('sets the image source correctly', () => {
        const image = page.getByTestId('prodcard-image-testid');
        expect(image.props.source).toEqual({ uri: inventoryMock.image });
      });
});