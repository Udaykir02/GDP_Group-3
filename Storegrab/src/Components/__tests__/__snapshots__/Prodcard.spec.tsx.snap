// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prodcard should render correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      style={
        [
          {
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          {
            "flex": 1,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              [
                {
                  "backgroundColor": "#fff",
                  "borderWidth": 0,
                  "marginBottom": 16,
                  "marginVertical": 16,
                  "minHeight": 114,
                },
                {
                  "paddingHorizontal": 16,
                  "paddingVertical": 8,
                },
              ],
              {
                "flexDirection": "row",
              },
            ]
          }
        >
          <View
            style={
              [
                [
                  {
                    "borderRadius": 3,
                    "elevation": 1,
                    "overflow": "hidden",
                  },
                  {
                    "borderBottomRightRadius": 0,
                    "borderTopRightRadius": 0,
                  },
                  {
                    "elevation": 2,
                    "shadowColor": "#000",
                    "shadowOffset": {
                      "height": 2,
                      "width": 0,
                    },
                    "shadowOpacity": 0.1,
                    "shadowRadius": 4,
                  },
                ],
                {
                  "flex": 0.3,
                },
              ]
            }
          >
            <Image
              source={
                {
                  "uri": "https://fastly.picsum.photos/id/237/200/300.jpg?hmac=TmmQSbShHz9CdQm0NkEjx1Dyh_Y984R9LpNrpvH2D_U",
                }
              }
              style={
                [
                  {
                    "borderRadius": 3,
                    "height": 107,
                    "width": "auto",
                  },
                  {
                    "borderRadius": 10,
                  },
                ]
              }
              testID="prodcard-image-testid"
            />
          </View>
          <View
            style={
              {
                "flex": 0.7,
                "justifyContent": "space-between",
              }
            }
          >
            <View
              style={
                {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                }
              }
            >
              <View
                style={
                  [
                    {
                      "flex": 0.4,
                    },
                    {
                      "marginHorizontal": 5,
                    },
                  ]
                }
              >
                <Text
                  style={
                    [
                      {
                        "flex": 1,
                        "flexWrap": "wrap",
                        "paddingBottom": 6,
                      },
                      {
                        "color": "#32325D",
                        "fontSize": 14,
                        "fontWeight": "bold",
                      },
                    ]
                  }
                >
                  item
                </Text>
                <Text
                  style={
                    [
                      {
                        "flex": 1,
                        "flexWrap": "wrap",
                        "paddingBottom": 6,
                      },
                      {
                        "color": "#32325D",
                        "fontSize": 12,
                        "fontWeight": "normal",
                      },
                    ]
                  }
                >
                  red
                </Text>
              </View>
              <View
                style={
                  {
                    "alignItems": "center",
                    "flex": 0.6,
                    "justifyContent": "center",
                  }
                }
              >
                <Text
                  style={
                    [
                      {
                        "flex": 1,
                        "flexWrap": "wrap",
                        "paddingBottom": 6,
                      },
                      {
                        "color": "#32325D",
                        "fontSize": 18,
                        "fontWeight": "bold",
                      },
                    ]
                  }
                >
                  $100
                </Text>
                <View
                  style={
                    [
                      {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "height": 40,
                        "justifyContent": "space-evenly",
                        "padding": 4,
                      },
                    ]
                  }
                >
                  <View
                    style={
                      {
                        "alignItems": "center",
                        "backgroundColor": "#ffffff",
                        "blurRadius": 8,
                        "borderColor": "#558235",
                        "borderRadius": 50,
                        "borderWidth": 1,
                        "flexDirection": "row",
                        "height": 40,
                        "justifyContent": "center",
                        "padding": 4,
                        "shadowColor": "transparent",
                        "shadowOffset": {
                          "height": 0,
                          "width": 0,
                        },
                        "shadowOpacity": 0.8,
                        "shadowRadius": 2,
                      }
                    }
                  >
                    <View
                      style={
                        {
                          "alignItems": "center",
                          "backgroundColor": "#ffffff",
                          "borderColor": "#6aa342",
                          "borderRadius": 50,
                          "borderWidth": 1,
                          "flexDirection": "row",
                          "height": 32,
                          "justifyContent": "center",
                          "padding": 4,
                        }
                      }
                    >
                      <View
                        accessibilityState={
                          {
                            "busy": undefined,
                            "checked": undefined,
                            "disabled": undefined,
                            "expanded": undefined,
                            "selected": undefined,
                          }
                        }
                        accessibilityValue={
                          {
                            "max": undefined,
                            "min": undefined,
                            "now": undefined,
                            "text": undefined,
                          }
                        }
                        accessible={true}
                        collapsable={false}
                        focusable={true}
                        onClick={[Function]}
                        onResponderGrant={[Function]}
                        onResponderMove={[Function]}
                        onResponderRelease={[Function]}
                        onResponderTerminate={[Function]}
                        onResponderTerminationRequest={[Function]}
                        onStartShouldSetResponder={[Function]}
                        style={
                          {
                            "opacity": 1,
                          }
                        }
                      >
                        <View
                          style={
                            {
                              "alignItems": "center",
                              "flexDirection": "column",
                              "justifyContent": "center",
                              "minWidth": 32,
                            }
                          }
                        >
                          <Text
                            style={
                              {
                                "color": "#000",
                                "fontSize": 16,
                                "fontWeight": "bold",
                                "height": 20,
                                "marginBottom": 0,
                                "marginHorizontal": 4,
                                "minWidth": 32,
                                "textAlign": "center",
                              }
                            }
                          >
                            Add
                          </Text>
                        </View>
                      </View>
                      <View
                        accessibilityState={
                          {
                            "busy": undefined,
                            "checked": undefined,
                            "disabled": undefined,
                            "expanded": undefined,
                            "selected": undefined,
                          }
                        }
                        accessibilityValue={
                          {
                            "max": undefined,
                            "min": undefined,
                            "now": undefined,
                            "text": undefined,
                          }
                        }
                        accessible={true}
                        collapsable={false}
                        focusable={true}
                        onClick={[Function]}
                        onResponderGrant={[Function]}
                        onResponderMove={[Function]}
                        onResponderRelease={[Function]}
                        onResponderTerminate={[Function]}
                        onResponderTerminationRequest={[Function]}
                        onStartShouldSetResponder={[Function]}
                        style={
                          {
                            "opacity": 1,
                          }
                        }
                      >
                        <View
                          collapsable={false}
                          style={
                            {
                              "alignItems": "center",
                              "backgroundColor": "rgba(255, 255, 255, 0)",
                              "borderColor": "#000000",
                              "borderRadius": 16,
                              "borderStyle": "dotted",
                              "borderWidth": 0,
                              "color": "#558235",
                              "height": 32,
                              "justifyContent": "center",
                              "marginHorizontal": 5,
                              "width": 32,
                            }
                          }
                        >
                          <Text
                            style={
                              {
                                "color": "#558235",
                                "fontSize": 16,
                              }
                            }
                          >
                            <Text
                              allowFontScaling={false}
                              selectable={false}
                              style={
                                [
                                  {
                                    "color": undefined,
                                    "fontSize": 32,
                                  },
                                  undefined,
                                  {
                                    "fontFamily": "Material Design Icons",
                                    "fontStyle": "normal",
                                    "fontWeight": "normal",
                                  },
                                  {},
                                ]
                              }
                            >
                              󰐕
                            </Text>
                          </Text>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
                <Text
                  style={
                    [
                      {
                        "flex": 1,
                        "flexWrap": "wrap",
                        "paddingBottom": 6,
                      },
                      {
                        "color": "#32325D",
                        "fontSize": 12,
                        "fontWeight": "bold",
                        "marginTop": 10,
                      },
                    ]
                  }
                >
                  In Stock: 
                  1
                </Text>
              </View>
            </View>
            <View
              style={
                {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "margin": 5,
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "flex": 1,
                      "flexWrap": "wrap",
                      "paddingBottom": 6,
                    },
                    {
                      "color": "#32325D",
                      "fontSize": 12,
                      "fontWeight": "bold",
                    },
                  ]
                }
              >
                Description: 
                description
              </Text>
            </View>
            <View
              style={
                [
                  {
                    "justifyContent": "space-between",
                  },
                  {
                    "marginHorizontal": 5,
                  },
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "flex": 1,
                      "flexWrap": "wrap",
                      "paddingBottom": 6,
                    },
                    {
                      "color": "#32325D",
                      "fontSize": 12,
                      "fontWeight": "bold",
                    },
                  ]
                }
              >
                brand
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
